src/smpi/internals/smpi_replay.cpp

```diff
@@ -373,24 +464,29 @@ static void action_test(const char *const *action){
     int flag = Request::test(&request, &status);
 
     XBT_DEBUG("MPI_Test result: %d", flag);
-    /* push back request in vector to be caught by a subsequent wait. if the test did succeed, the request is now nullptr.*/
-    get_reqq_self()->push_back(request);
 
+    register_request_with_key(request, key);
+  
     TRACE_smpi_testing_out(rank);
   }
   log_timed_action (action, clock);
 }
 
 static void action_wait(const char *const *action){
-  CHECK_ACTION_PARAMS(action, 0, 0)
+  CHECK_ACTION_PARAMS(action, 3, 0);
+  int src = atoi(action[2]);
+  int dst = atoi(action[3]);
+  int tag = atoi(action[4]);
   double clock = smpi_process()->simulated_elapsed();
+  std::string key = build_request_key(src, dst, tag);
+  MPI_Request request = get_request_with_key(key);
```

Comment:
This is probably the reason why the request is stored with a string as the key: In this case, we don't seem to have any request and so it's impossible to easily retrieve the request? However, the old version just accessed the last element (see the diff). Why is this not true any longer?